src/Jackalope/Version/VersionHandler.php

```diff
@@ -246,4 +246,60 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
             $frozenNode->setProperty($propertyName, $property->getValue());
         }
     }
+
+    public function restoreItem($removeExisting, $versionPath, $path)
+    {
+        $node = $this->objectManager->getNodeByPath($path);
+
+        if ($node->isModified()) {
+            throw new InvalidItemStateException(
+                sprintf(
+                    'Node "%s" contains unsaved changes',
+                    $path
+                )
+            );
+        }
+
+        $versionNode = $this->objectManager->getNodeByPath($versionPath, 'Version\Version');
+        $frozenNode = $versionNode->getNode('jcr:frozenNode');
+
+        // TODO reset primary type
+
+        if ($frozenNode->hasProperty('jcr:frozenMixinTypes')) {
+            $node->setProperty('jcr:mixinTypes', $frozenNode->getPropertyValue('jcr:frozenMixinTypes'));
+        }
```

Comment:
is that property not always set? if not, you would need to delete the mixinTypes from the node in case it has some.